homeassistant/components/tolo/config_flow.py

```diff
@@ -37,57 +49,70 @@ def _check_device_availability(host: str) -> bool:
+                    # Host didn't change, just reload with same data
+                    return self.async_update_reload_and_abort(
+                        self._get_reconfigure_entry(), data_updates=user_input
+                    )
 
             device_available = await self.hass.async_add_executor_job(
                 self._check_device_availability, user_input[CONF_HOST]
             )
 
-            if not device_available:
-                errors["base"] = "cannot_connect"
-            else:
-                return self.async_create_entry(
-                    title=DEFAULT_NAME, data={CONF_HOST: user_input[CONF_HOST]}
-                )
+            if device_available:
+                if self.source == SOURCE_RECONFIGURE:
+                    return self.async_update_reload_and_abort(
+                        self._get_reconfigure_entry(), data_updates=user_input
+                    )
+                return self.async_create_entry(title=DEFAULT_NAME, data=user_input)
+
+            errors["base"] = "cannot_connect"
+
+        schema_values: dict[str, Any] | MappingProxyType[str, Any] = {}
+        if user_input is not None:
+            schema_values = user_input
+        elif self.source == SOURCE_RECONFIGURE:
+            schema_values = self._get_reconfigure_entry().data
+        elif self.source == SOURCE_DHCP and self._dhcp_discovery_info is not None:
+            schema_values = {CONF_HOST: self._dhcp_discovery_info.ip}
 
         return self.async_show_form(
             step_id="user",
-            data_schema=vol.Schema({vol.Required(CONF_HOST): str}),
+            data_schema=self.add_suggested_values_to_schema(
+                CONFIG_SCHEMA,
+                schema_values,
+            ),
             errors=errors,
         )
 
+    async def async_step_reconfigure(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle a reconfiguration config flow initialized by the user."""
+        return await self.async_step_user(user_input)
+
     async def async_step_dhcp(
         self, discovery_info: DhcpServiceInfo
     ) -> ConfigFlowResult:
         """Handle a flow initialized by discovery."""
+        # Check for existing entries with the same host first
+        self._async_abort_entries_match({CONF_HOST: discovery_info.ip})
+
         await self.async_set_unique_id(format_mac(discovery_info.macaddress))
         self._abort_if_unique_id_configured({CONF_HOST: discovery_info.ip})
-        self._async_abort_entries_match({CONF_HOST: discovery_info.ip})
 
         device_available = await self.hass.async_add_executor_job(
             self._check_device_availability, discovery_info.ip
         )
 
         if device_available:
-            self._discovered_host = discovery_info.ip
-            return await self.async_step_confirm()
+            self._dhcp_discovery_info = discovery_info
+            return await self.async_step_user()
```

Comment:
I think it's weird to redirect to the user flow here, it means we will ask for an IP-address for DHCP-flows?